src/common/libflux/test/mpipe.c

```diff
@@ -0,0 +1,281 @@
+    flux_msg_t *msg;
+    struct pollfd pfd;
+
+    if (!(msg1 = flux_request_encode ("foo", NULL)))
+        BAIL_OUT ("flux_request_encode failed");
+    if (!(msg2 = flux_request_encode ("foo", NULL)))
+        BAIL_OUT ("flux_request_encode failed");
+
+    ok ((q = mpipe_create (0)) != NULL,
+        "mpipe_create works");
+    ok (mpipe_pollevents (q) == POLLOUT,
+        "mpipe_pollevents on empty queue returns POLLOUT");
+    ok (mpipe_enqueue (q, &msg1) == 0,
+        "mpipe_enqueue msg1 works");
+    ok (mpipe_pollevents (q) == (POLLOUT | POLLIN),
+        "mpipe_pollevents on non-empty queue returns POLLOUT|POLLIN");
+    ok (mpipe_enqueue (q, &msg2) == 0,
+        "mpipe_enqueue msg2 works");
+    ok (mpipe_pollevents (q) == (POLLOUT | POLLIN),
+        "mpipe_pollevents still returns POLLOUT|POLLIN");
+    ok ((msg = mpipe_dequeue (q)) != NULL,
+        "mpipe_dequeue returns a message");
+    flux_msg_decref (msg);
+    ok (mpipe_pollevents (q) == (POLLOUT | POLLIN),
+        "mpipe_pollevents still returns POLLOUT|POLLIN");
+
+    ok ((msg = mpipe_dequeue (q)) != NULL,
+        "mpipe_dequeue returns a message");
+    flux_msg_decref (msg);
+    ok (mpipe_pollevents (q) == POLLOUT,
+        "mpipe_pollevents on empty queue returns POLLOUT");
+
+    /* now test pollfd */
+    if (!(msg1 = flux_request_encode ("foo", NULL)))
+        BAIL_OUT ("flux_request_encode failed");
+
+    ok ((pfd.fd = mpipe_pollfd (q)) >= 0,
+        "mpipe_pollfd works");
+    pfd.events = POLLIN,
+    pfd.revents = 0,
+    ok (poll (&pfd, 1, 0) == 1 && pfd.revents == POLLIN,
+        "mpipe_pollfd suggests we read pollevents");
+    ok (mpipe_pollevents (q) == POLLOUT,
+        "mpipe_pollevents on empty queue returns POLLOUT");
+    pfd.events = POLLIN,
+    pfd.revents = 0,
+    ok (poll (&pfd, 1, 0) == 0,
+        "pollfd is no longer ready");
+    ok (mpipe_enqueue (q, &msg1) == 0,
+        "mpipe_enqueue works");
+    pfd.events = POLLIN,
+    pfd.revents = 0,
+    ok (poll (&pfd, 1, 0) == 1 && pfd.revents == POLLIN,
+        "pollfd suggests we read pollevents");
+    ok (mpipe_pollevents (q) == (POLLOUT | POLLIN),
+        "mpipe_pollevents on non-empty queue returns POLLOUT|POLLIN");
+    pfd.events = POLLIN,
+    pfd.revents = 0,
+    ok (poll (&pfd, 1, 0) == 0,
+        "pollfd is no longer ready");
+    ok (mpipe_pollevents (q) == (POLLOUT | POLLIN),
+        "mpipe_pollevents still returns POLLOUT|POLLIN");
+
+    mpipe_destroy (q);
+}
+
+void check_single_thread (void)
+{
+    struct mpipe *q;
+    flux_msg_t *msg1;
+    void *msg1_ptr;
+    flux_msg_t *msg;
+
+    if (!(msg1_ptr = msg1 = flux_msg_create (FLUX_MSGTYPE_REQUEST)))
+        BAIL_OUT ("flux_msg_create failed");
+
+    q = mpipe_create (MPIPE_SINGLE_THREAD);
+    ok (q != NULL,
+        "mpipe_create flags=SINGLE_THREAD works");
+    flux_msg_incref (msg1);
+    ok (mpipe_enqueue (q, &msg1) == 0,
+        "mpipe_enqueue msg1 works with refcount==2");
+    flux_msg_decref (msg1);
```

Comment:
valgrind detects a leak of `msg1` here:
```
==4088027== 120 bytes in 1 blocks are definitely lost in loss record 1 of 1
==4088027==    at 0x4C3CE4B: calloc (vg_replace_malloc.c:1328)
==4088027==    by 0x402AA9: flux_msg_create (message.c:91)
==4088027==    by 0x401DC1: check_single_thread (mpipe.c:170)
==4088027==    by 0x401435: main (mpipe.c:273)
```

When peeking at the code I realized I'm not quite sure how the `mpipe` interface works. Does `mpipe_enqueue` steal the reference to `msg`? But maybe not when `MPIPE_SINGLE_THREAD` is used? Anyway, it is probably obvious, but still maybe the header file could use a couple comments? (And I'm not certain how to fix the leak here, probably something obvious)

Reply:
Ah, maybe `mpipe_enqueue()` is setting `*msg = NULL` here so the `flux_msg_decref()` has no effect?

Reply:
```suggestion
    flux_msg_incref (msg1);
    ok (mpipe_enqueue (q, &msg1) == 0,
        "mpipe_enqueue msg1 works with refcount==2");
    flux_msg_decref (msg1_ptr);
```

Reply:
Yes, that's it, good catch.

I'll add some documentation to the header.  Also, now I'm wondering if should use the zeromq destructor style here given that messages with external references are accepted in single thread mode.   It made more sense before that feature was added.